### PR TITLE
Add option to run Dev-env with FakeIPA

### PR DESCRIPTION
**Pull request by mboukhalfa**
Run environment to test scalability with FakeIPA and also download FKAS image

To enable fakeIPA in devenv add the following line to the config file
```
export NODES_PLATFORM="fake"
```

Tested from : https://github.com/metal3-io/cluster-api-provider-metal3/pull/2010

### CODE REVIEW

File: 03_launch_mgmt_cluster.sh
Comment (tuminoid) [inline 03_launch_mgmt_cluster.sh:546]:
```suggestion
    kubectl -n legacy get secret -n baremetal-operator-system ironic-cert -o json -o=jsonpath="{.data.ca\.crt}" | base64 -d > /opt/metal3-dev-env/fake-ipa/ironic-ca.crt
```

File: 03_launch_mgmt_cluster.sh
Comment (tuminoid) [inline 03_launch_mgmt_cluster.sh:582]:
```suggestion
    environment.metal3.io: baremetal
```

File: 03_launch_mgmt_cluster.sh
Comment (tuminoid) [inline 03_launch_mgmt_cluster.sh:602]:
```suggestion
    apply_bm_hosts "$NAMESPACE"
```

File: 03_launch_mgmt_cluster.sh
Comment (tuminoid) [inline 03_launch_mgmt_cluster.sh:541]:
```suggestion
  # if fake platform (no VMs) run FakeIPA
```

File: 03_launch_mgmt_cluster.sh
Comment (tuminoid) [inline 03_launch_mgmt_cluster.sh:559]:
```suggestion
    environment.metal3.io: baremetal
```

File: 03_launch_mgmt_cluster.sh
Comment (tuminoid) [inline 03_launch_mgmt_cluster.sh:600]:
```suggestion
    kubectl apply -f "${WORKING_DIR}/bmhosts_crs.yaml" -n "$NAMESPACE"
```

File: 03_launch_mgmt_cluster.sh
Comment (tuminoid) [inline 03_launch_mgmt_cluster.sh:606]:
```suggestion
  source tilt-setup/deploy_tilt_env.sh
```

File: 03_launch_mgmt_cluster.sh
Comment (tuminoid) [inline 03_launch_mgmt_cluster.sh:605]:
```suggestion
```

**Review by tuminoid — COMMENTED** [review]
Some indentation issues.

File: 03_launch_mgmt_cluster.sh
Comment (tuminoid) [inline 03_launch_mgmt_cluster.sh:184]:
```suggestion
  if [[ "${VM_PLATFORM}" == "fake" ]]; then
```

**Review by tuminoid — COMMENTED** [review]
Also, add the new environment variables to `vars.md`, same as sushy-tools vars are there.

**Comment by mboukhalfa** [comment]
/hold 
Adding FKAS WIP

**Comment by mboukhalfa** [comment]
> /hold Adding FKAS WIP

We will add FKAS in the e2e side 

/hold cancel

**Comment by mboukhalfa** [comment]
> Run environment to test scalability with FakeIPA
> 
> To enable fakeIPA in devenv add the following line to the config file
> 
> ```
> export vm_platform="fake"
> ```
> 
> Enhancement TODO:
> 
>     * Skip install package that are not used by fake env like libvirt
> 
>     * Skip downloading real images

This will be tested from : https://github.com/metal3-io/cluster-api-provider-metal3/pull/2010

File: 01_prepare_host.sh
Comment (mquhuy) [inline 01_prepare_host.sh:146]: Could you edit this log to give a bit of context, like "Skipping image prepulling on fake nodes platform"?

File: 04_verify.sh
Comment (mquhuy) [inline 04_verify.sh:23]: echo "Skipping virsh nodes verification on fake vm platform"

File: vm-setup/roles/libvirt/tasks/network_setup_tasks.yml
Comment (mquhuy) [inline vm-setup/roles/libvirt/tasks/network_setup_tasks.yml:20]: Not sure why this needs to be splitted to a new file?

**Review by mquhuy — COMMENTED** [review]
Some nits

**Comment by mquhuy** [comment]
LGTM

**Review by smoshiur1237 — COMMENTED** [review]
Thanks @mboukhalfa for adding the FakeIPA option with dev-env. I would request others to review it. After the review I can add approval.
LGTM

File: 01_prepare_host.sh
Comment (Rozzii) [inline 01_prepare_host.sh:146]: I would prefer this string to be quoted.

File: 02_configure_host.sh
Comment (Rozzii) [inline 02_configure_host.sh:412]: quote all of these please e.g.
`FAKE_IPA_IMAGE="${FAKE_IPA_LOCAL_IMAGE:-${FAKE_IPA_IMAGE}}"`

File: 03_launch_mgmt_cluster.sh
Comment (Rozzii) [inline 03_launch_mgmt_cluster.sh:234]:
```suggestion
 if [[ "${EPHEMERAL_CLUSTER}" == "kind" && "${IRONIC_TLS_SETUP}" == "true" ]]; then
```

Comment (Rozzii) [reply 03_launch_mgmt_cluster.sh:234]: Do it only if it fits the 80 char limit.

File: 03_launch_mgmt_cluster.sh
Comment (kashifest) [inline 03_launch_mgmt_cluster.sh:258]: I am thinking if we should add this function in a separate script file for better management and keeping 03_ less consumed?

**Review by kashifest — COMMENTED** [review]
small nits, otherwise lgtm

File: 03_launch_mgmt_cluster.sh
Comment (mboukhalfa) [reply 03_launch_mgmt_cluster.sh:178]: yes that is intentional but we can remove this since if https://github.com/metal3-io/utility-images/pull/24 merged

File: 03_launch_mgmt_cluster.sh
Comment (mboukhalfa) [reply 03_launch_mgmt_cluster.sh:258]: which function you mean ? this container run ?

File: 03_launch_mgmt_cluster.sh
Comment (mboukhalfa) [reply 03_launch_mgmt_cluster.sh:234]: Hmm but the else here is only for the first condition ?  I do not think it will same isn't it ?

Comment (Rozzii) [reply 03_launch_mgmt_cluster.sh:234]: in the outer if you don't have anything else right now other than the inner "inner" if so it wouldn't change anything in practice.

Comment (Rozzii) [reply 03_launch_mgmt_cluster.sh:234]: Is the no-op intentional when "${EPHEMERAL_CLUSTER}" == "kind" and "${IRONIC_TLS_SETUP}" == "false"?
If yes then I am fine leaving it like this.

**Review by Rozzii — CHANGES_REQUESTED** [review]
Most of the issues have been resolved, I still have ongoing question for one of the topics.

File: 03_launch_mgmt_cluster.sh
Comment (kashifest) [reply 03_launch_mgmt_cluster.sh:256]: `launch_fakeIPA()`

**Comment by mboukhalfa** [comment]
> Most of the issues have been resolved, I still have ongoing question for one of the topics.

Addressed

File: 03_launch_mgmt_cluster.sh
Comment (mboukhalfa) [reply 03_launch_mgmt_cluster.sh:256]: I am just following the old structure all the launch function are defined here ..

Comment (kashifest) [reply 03_launch_mgmt_cluster.sh:256]: You are correct, indeed we have quite a few of those launch functions. However, most of the launch functions are indeed a part of the 03 normal workflow (they are almost always called) whereas I feel this fake_ipa function is not. That being said, this is not a functional requirement. So I am ok if you insist to keep it here.

File: 03_launch_mgmt_cluster.sh
Comment (lentzi90) [inline 03_launch_mgmt_cluster.sh:239]: Typo?
```suggestion
    kubectl get secret -n baremetal-operator-system ironic-cert -o json -o=jsonpath="{.data.ca\.crt}" | base64 -d > /opt/metal3-dev-env/fake-ipa/ironic-ca.crt
```

File: 04_verify.sh
Comment (lentzi90) [inline 04_verify.sh:24]: Can we move this down to line ~260? Then we will still verify that the kubernetes cluster is there and has some resources.

File: vm-setup/roles/common/tasks/write_ironic_nodes_tasks.yml
Comment (lentzi90) [inline vm-setup/roles/common/tasks/write_ironic_nodes_tasks.yml:5]: nit: This comment makes sense where the task is used but not here where it is defined.

File: 03_launch_mgmt_cluster.sh
Comment (tuminoid) [inline 03_launch_mgmt_cluster.sh:247]: Why under `/root` ?

File: 03_launch_mgmt_cluster.sh
Comment (tuminoid) [inline 03_launch_mgmt_cluster.sh:582]:
```suggestion
    list_nodes | make_bm_hosts
```

File: 03_launch_mgmt_cluster.sh
Comment (tuminoid) [inline 03_launch_mgmt_cluster.sh:585]:
```suggestion
    apply_bm_hosts "${NAMESPACE}"
```

File: 03_launch_mgmt_cluster.sh
Comment (tuminoid) [inline 03_launch_mgmt_cluster.sh:592]:
```suggestion
  . tilt-setup/deploy_tilt_env.sh
```

**Review by tuminoid — COMMENTED** [review]
Some minor nits.

File: 04_verify.sh
Comment (mboukhalfa) [reply 04_verify.sh:24]: I moved it to 240 because with the fake we do not apply BMH we do need to verify them

File: 03_launch_mgmt_cluster.sh
Comment (mboukhalfa) [reply 03_launch_mgmt_cluster.sh:239]: nice catch cool that it was still working 😆

File: 03_launch_mgmt_cluster.sh
Comment (mboukhalfa) [reply 03_launch_mgmt_cluster.sh:247]: Just following the current structure of vbmc and sushytools

Comment (lentzi90) [reply 03_launch_mgmt_cluster.sh:247]: To add to that, I think the current structure is because these are used inside containers and those containers must likely use the root user.

**Review by lentzi90 — COMMENTED** [review]
/lgtm

**Pull request merged**
